File: javascript/after-refactor/Args.js
const { ArgsError, ErrorCode } = require('./ArgsError');
const { BooleanArgumentMarshaler } = require('./BooleanArgumentMarshaler');
const { IntegerArgumentMarshaler } = require('./IntegerArgumentMarshaler');
const { DoubleArgumentMarshaler } = require('./DoubleArgumentMarshaler');
const { StringArgumentMarshaler } = require('./StringArgumentMarshaler');

class Args {
  constructor(schema, args) {
    this.marshalers = new Map();
    this.currentArgument = undefined;
    this.parseSchema(schema);
    this.parseArgumentStrings(args);
  }

  parseSchema(schema) {
    schema.split(',')
      .filter((element) => element.length > 0)
      .forEach((element) => this.parseSchemaElement(element.trim()));
  }

  parseSchemaElement(element) {
    const elementID = element[0];
    const elementTail = element.substring(1);
    this.validateSchemaElementId(elementID);
    if (elementTail.length === 0) {
      this.marshalers.set(elementID, new BooleanArgumentMarshaler());
    } else if (elementTail === '*') {
      this.marshalers.set(elementID, new StringArgumentMarshaler());
    } else if (elementTail === '#') {
      this.marshalers.set(elementID, new IntegerArgumentMarshaler());
    } else if (elementTail === '##') {
      this.marshalers.set(elementID, new DoubleArgumentMarshaler());
    } else {
      throw new ArgsError(ErrorCode.INVALID_ARGUMENT_FORMAT, elementID, elementTail);
    }
  }

  validateSchemaElementId(elementID) {
    if (elementID.length !== 1 || !elementID.match(/[a-z]/i)) {
      throw new ArgsError(ErrorCode.INVALID_ARGUMENT_NAME, elementID);
    }
  }

  parseArgumentStrings(args) {
    if (args.length === 0) {
      return;
    }
    this.currentArgument = args[Symbol.iterator]();
    for (;;) {
      const { value: argString, done } = this.currentArgument.next();
      if (done) {
        break;
      }
      if (argString.startsWith('-')) {
        this.parseArgumentCharacters(argString.substring(1));
      } else {
        throw new ArgsError(ErrorCode.INVALID_ARGUMENT_FORMAT, undefined, argString);
      }
    }
  }

  parseArgumentCharacters(argChars) {
    for (let i = 0; i < argChars.length; i += 1) {
      this.parseArgumentCharacter(argChars[i]);
    }
  }

  parseArgumentCharacter(argChar) {
    const m = this.marshalers.get(argChar);
    if (!m) {
      throw new ArgsError(ErrorCode.UNEXPECTED_ARGUMENT, argChar);
    } else {
      try {
        m.set(this.currentArgument);
      } catch (e) {
        e.setErrorArgumentID(argChar);
        throw e;
      }
    }
  }

  getBoolean(arg) {
    return BooleanArgumentMarshaler.getValue(this.marshalers.get(arg));
  }

  getString(arg) {
    return StringArgumentMarshaler.getValue(this.marshalers.get(arg));
  }

  getInt(arg) {
    return IntegerArgumentMarshaler.getValue(this.marshalers.get(arg));
  }

  getDouble(arg) {
    return DoubleArgumentMarshaler.getValue(this.marshalers.get(arg));
  }
}

module.exports = {
  Args,
  ErrorCode,
  ArgsError,
};

File: java/src/main/java/args/after/Args.java
package args.after;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;

/**
 * Args parses CLI args.
 */
public class Args {
  private Map<Character, ArgumentMarshaler> marshalers;
  private ListIterator<String> currentArgument;

  /**
   * Args Constructor.
   *
   * @param schema Schema of the args, see project README for more details
   * @param args   CLI args to parse
   * @throws ArgsException if there is a problem parsing the args
   */
  public Args(String schema, String[] args) throws ArgsException {
    marshalers = new HashMap<Character, ArgumentMarshaler>();
    parseSchema(schema);
    parseArgumentStrings(Arrays.asList(args));
  }

  private void parseSchema(String schema) throws ArgsException {
    for (String element : schema.split(",")) {
      if (element.length() > 0) {
        parseSchemaElement(element.trim());
      }
    }
  }

  private void parseSchemaElement(String element) throws ArgsException {
    char elementId = element.charAt(0);
    String elementTail = element.substring(1);
    validateSchemaElementId(elementId);
    if (elementTail.length() == 0) {
      marshalers.put(elementId, new BooleanArgumentMarshaler());
    } else if (elementTail.equals("*")) {
      marshalers.put(elementId, new StringArgumentMarshaler());
    } else if (elementTail.equals("#")) {
      marshalers.put(elementId, new IntegerArgumentMarshaler());
    } else if (elementTail.equals("##")) {
      marshalers.put(elementId, new DoubleArgumentMarshaler());
    } else {
      throw new ArgsException(ArgsErrorCode.INVALID_ARGUMENT_FORMAT, elementId, elementTail);
    }
  }

  private void validateSchemaElementId(char elementId) throws ArgsException {
    if (!Character.isLetter(elementId)) {
      throw new ArgsException(ArgsErrorCode.INVALID_ARGUMENT_NAME, elementId, null);
    }
  }

  private void parseArgumentStrings(List<String> argsList) throws ArgsException {
    if (argsList.size() == 0) {
      return;
    }
    currentArgument = argsList.listIterator();
    while (currentArgument.hasNext()) {
      String argString = currentArgument.next();
      if (!argString.startsWith("-")) {
        throw new ArgsException(ArgsErrorCode.INVALID_ARGUMENT_FORMAT, argString);
      }
      parseArgumentCharacters(argString.substring(1));
    }
  }

  private void parseArgumentCharacters(String argChars) throws ArgsException {
    for (int i = 0; i < argChars.length(); i++) {
      parseArgumentCharacter(argChars.charAt(i));
    }
  }

  private void parseArgumentCharacter(char argChar) throws ArgsException {
    ArgumentMarshaler m = marshalers.get(argChar);
    if (m == null) {
      throw new ArgsException(ArgsErrorCode.UNEXPECTED_ARGUMENT, argChar, null);
    }
    try {
      m.set(currentArgument);
    } catch (ArgsException e) {
      e.setErrorArgumentId(argChar);
      throw e;
    }
  }

  public boolean getBoolean(char arg) {
    return BooleanArgumentMarshaler.getValue(marshalers.get(arg));
  }

  public String getString(char arg) {
    return StringArgumentMarshaler.getValue(marshalers.get(arg));
  }

  public int getInt(char arg) {
    return IntegerArgumentMarshaler.getValue(marshalers.get(arg));
  }

  public double getDouble(char arg) {
    return DoubleArgumentMarshaler.getValue(marshalers.get(arg));
  }
}
